mev-build-rs/src/mempool_builder/builder/mod.rs

```diff
@@ -0,0 +1,472 @@
+pub struct BuildContext {
+    pub slot: Slot,
+    pub parent_hash: H256,
+    pub proposer: BlsPublicKey,
+    pub timestamp: u64,
+    pub proposer_fee_recipient: ExecutionAddress,
+    pub prev_randao: Hash32,
+    pub withdrawals: Vec<Withdrawal>,
+    pub relays: Vec<RelayIndex>,
+    pub chain_spec: Arc<ChainSpec>,
+    pub block_env: BlockEnv,
+    pub cfg_env: CfgEnv,
+    pub extra_data: Bytes,
+    pub parent_block: Arc<SealedBlock>,
+    pub builder_wallet: LocalWallet,
+    // Amount of gas to reserve after building a payload
+    // e.g. used for end-of-block proposer payments
+    pub gas_reserve: u64,
+    // Amount of the block's value to bid to the proposer
+    pub bid_percent: f64,
+    // Amount to add to the block's value to bid to the proposer
+    pub subsidy: U256,
+}
+
+impl BuildContext {
+    pub fn id(&self) -> BuildIdentifier {
+        let mut data = Vec::with_capacity(88);
+        self.slot.serialize(&mut data).expect("can serialize");
+        self.parent_hash.serialize(&mut data).expect("can serialize");
+        self.proposer.serialize(&mut data).expect("can serialize");
+        hash(data)
+    }
+
+    pub fn base_fee(&self) -> u64 {
+        self.block_env.basefee.to::<u64>()
+    }
+
+    pub fn number(&self) -> u64 {
+        self.block_env.number.to::<u64>()
+    }
+
+    pub fn gas_limit(&self) -> u64 {
+        self.block_env.gas_limit.try_into().unwrap_or(u64::MAX)
+    }
+}
+
+#[derive(Default, Debug)]
+struct State {
+    proposer_schedule:
+        BTreeMap<Slot, HashMap<BlsPublicKey, HashMap<ValidatorPreferences, Vec<RelayIndex>>>>,
+    jobs: HashMap<BuildIdentifier, (Arc<BuildContext>, Arc<PayloadWithFees>)>,
+}
+
+impl<
+        Pool: reth_transaction_pool::TransactionPool,
+        Provider: reth_provider::StateProviderFactory + reth_provider::BlockReaderIdExt,
+    > Builder<Pool, Provider>
+{
+    pub fn new(
+        secret_key: SecretKey,
+        context: Arc<Context>,
+        payload_attributes_tx: Sender<PayloadBuilderAttributes>,
+        relays: Vec<Relay>,
+        pool: Pool,
+        provider: Provider,
+        chain_spec: Arc<ChainSpec>,
+        extra_data: Bytes,
+        builder_wallet: LocalWallet,
+    ) -> Self {
+        let public_key = secret_key.public_key();
+
+        Self(Arc::new(Inner {
+            secret_key,
+            public_key,
+            context,
+            payload_attributes_tx,
+            relays,
+            pool,
+            provider,
+            chain_spec,
+            builder_wallet,
+            extra_data,
+            state: Default::default(),
+        }))
+    }
+
+    fn process_validator_schedule_for_relay(
+        &self,
+        relay: RelayIndex,
+        schedule: &[ProposerSchedule],
+    ) {
+        tracing::trace!(schedule_count = ?schedule.len(), relay = ?relay);
+
+        let mut state = self.state.lock().unwrap();
+        for duty in schedule {
+            tracing::trace!(slot = ?duty.slot, "processing schedule");
```

Comment:
consider: would be nice to have status updates on when the next proposal is -- useful on networks w/ sparser adoption